compiler/rustc_typeck/src/collect/type_of.rs

```diff
@@ -483,15 +512,59 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                     .discr_type()
                     .to_ty(tcx),
 
+                Node::TraitRef(trait_ref @ &TraitRef {
+                  path, ..
+                }) if let Some((binding, seg)) =
+                  path
+                      .segments
+                      .iter()
+                      .filter_map(|seg| seg.args.map(|args| (args.bindings, seg)))
+                      .find_map(|(bindings, seg)| {
```

Comment:
same thing as earlier, just use `find_map` and do the `filter_map` operation inside of it.